Shard several of the most costly targets.

shumway opened the pull request:
## Proposed changes

We want to reduce build time for the CK library that is consumed by MIOpen.

Several of the build units (source files) take ~20 minutes to compile. To reduce compilation time, we split those files
out into multiple copies (build shards). There are three complications:

1.  The kernel type instantiations to be sharded are in std::tuple types. We don't to break up those tuples this finely. 
    The solution we propose is to create a utility metafunction `filter_tuple_by_modulo_t` that splits up the kernel
    type tuples using a stride and offset.
2. To avoid duplicating a lot of the instantiation code, template the instantiation functions on a shard and offset.
3. To call the code, wrap the templated instantiation functions in the original instantiation function, and use
    `extern template` to make sure we don't reinstantiate these templates from the header code.

We partially automated this processes with code generation, by writing a CMake function 
generate_sharded_instantiations to generate the source files for the shared instatiation functions and the calling 
function directly from CMake.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

## Discussion

I went with this solution because the strategy of breaking down tuples into smaller tuples wasn't going to
get us all the individual kernels. This was the best compromise I could find between vebosity and getting 
build performance. Now we have about 165 targets for this target, which shard well on 128 shards:

time ninja device_grouped_conv3d_fwd_instance

This reduces build time from this target from about 22 minutes to 11 minutes.

shumway committed:
Shard several of the most costly targets.

Introduces a filter_tuple_by_modulo to break up tuples.

Drops build time of target from 21 minutes to under 14 minutes with 64
build processes, or 11 minutes with 128 build processes.

time ninja -j 64 device_grouped_conv3d_fwd_instance

illsilin committed:
fix clang format

spolifroni-amd reviewed (commented):
doesn't look like there's anything for docs.

andriy-ca commented on include/ck/utility/filter_tuple.hpp:
```diff
@@ -0,0 +1,66 @@
+
+namespace ck::util {
+
+template <typename Tuple, std::size_t Stride, std::size_t Offset>
```
Using the standard library breaks GEN_RTC functionality. We must use CK own analogs instead of `std`.

Please ensure the proposed changes pass the GEN_RTC tests.

Update. I can see that the Codegen Tests on gfx90a are passing. These changes do not interact with the Codegen just yet. Nevertheless, I would eliminate the `std` namespace as much as possible.

shumway replied:
Thanks! This code is intended for instantiating templates for MIOpen, and likely won't be part of any CK codegen. I see your point, though, and I'll just change this to size_t. (If there is a better reference for "CK's own analogs" please let me know.)

I noticed that ck_tile source make use of standard headers. Is there some guidance for style or project rules around this? I looked on Confluence and also in the GitHub project and couldn't find any developer guidelines. Please let me know if there's anything I should read to help reviews move easier.

shumway replied:
Actually, the file for this code should probably stay exactly as it is. It's specifically designed around splitting std::tuple, which is existing code used for instantiating templates for MIOpen. It should never be used for codegen. Ideally all this code (this util and everything else I'm modifying in this PR) would be in MIOpen, but the way we have this setup right now (including the heavy use of std::tuple in CK), all the MIOpen instantiation code is inside CK. I'm going to leave this as is.

shumway committed:
Fix build errors in instantiation code.

I wasn't sure how to test the header-only instantiation code on my
initial commit. From Jenkins CI test results, I see that there is a
test target that depends on these headers:

ninja -j 128 test_grouped_convnd_fwd

This allowed me to test the build locally. I found three mistakes I
made, mostly related to early experiments on I tried on the code.
This was hard to find earlier because this PR is really too large.

I also discovered that there are five 2D convolution targets that now
dominate the compilation time. I will likely address those in a later
PR, rather than adding even more changes to this PR.

shumway committed:
Fix link errors from mismatched declarations.

Our pattern for instantiating MIOpen templates uses duplicate
declarations (instead of headers). This is fragile, and I didn't
notice that my last commit had a bunch of link errors. I fixed these
mistakes, and the bin/test_grouped_conv_fwd test target binary now links
correctly.

bartekxk reviewed (changes requested):
Hi @shumway , great job. Is it possible to generate these c++ files using some python script? At now it is impossible to check/review if each instance has been added correctly. It will be much easier to handle such python generator instead of 150 files. It will also be very useful to port such python script to generate c++ files for bwd data and bwd weight

shumway commented:
> Hi @shumway , great job. Is it possible to generate these c++ files using some python script? At now it is impossible to check/review if each instance has been added correctly. It will be much easier to handle such python generator instead of 150 files. It will also be very useful to port such python script to generate c++ files for bwd data and bwd weight

@bartekxk and I chatted, and I'll look at existing codegen inside the CK codebase to see how to set this up so that it matches what's already here.

shumway commented:
@bartekxk 's suggestion to use codegen looks like a good idea. I have started to implement this as a cmake function, using `configure_file`. This has the advantage that it's pure CMake and we'll get back to essentially no extra hand-written source files. The input template generalizes a lot, but it make take me another day to get all the codegen rules sorted out and replace everything.

shumway committed:
Migrate the design to a code-generation approach.

Use a CMake function with template files to generate the source files for the
intantiating the kerenels and to generate the calling function.

shumway commented:
This code is much simpler now, down to 35 files changed. I followed @bartekxk 's suggestion and converted to code
generation. WIth this simplification, I can now shard the 2D convolution targets that are limiting the build time.

shumway committed:
Shard the longest 2D convolution builds

Now that we have automated the shard instantiation, we can shard the 2D
convolution targets that take the longest to build. The target
test_grouped_conv2d_fwd now compiles in 15 minutes.

bartekxk reviewed (dismissed):
lgtm

bartekxk commented:
Could you just make sanity check and verify if test_grouped_convnd_fwd generates the same output on the develop and on your branch?

shumway committed:
Use PROJECT_SOURCE_DIR for submodule compatibility

I used CMAKE_SOURCE_DIR to refer to the top-level source directory in
the ShardInstantiation.cmake file, but this can cause issues with
git submodules.  Instead, we should use PROJECT_SOURCE_DIR to ensure
compatibility when this project is used as a submodule in another
project.

shumway committed:
Merge branch 'develop' into shumway/refactor_targets
